lib/pbench/server/filetree.py

```diff
@@ -299,18 +312,34 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
         unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
+        unpacked.mkdir(parents=True)
         self.unpacked = unpacked
+
+        try:
+            process = subprocess.run(
+                ["tar", "-xf", self.tarball_path],
+                cwd=self.unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path.name, str(exc)) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballUnpackError(
+                    self.tarball_path.name,
+                    f"tar exited with status {process.returncode}:  {process.stderr!r}",
```

Comment:
Is it weird that we're including the newline(s) from `stderr`? Sorry I didn't think of that last time through; always a risk of a re-review! 😆 There's also the (small?) risk that there'll be multiple lines, potentially long... though I don't think `tar` ever gets all that verbose. Should we maybe do a `.split("\n")` and include the array? Would that look even more ugly? 😕

Reply:
I don't see any benefit to _this_ code doing the `split()` -- the array is no easier to deal with than the original string with the embedded newlines, and the caller (er, `except`-er) can always do the split itself.  (I could see doing a `strip()` (to remove the _trailing_ newline), but I wouldn't do the `split()`.)

Reply:
I think it's weird to raise an exception with embedded newlines in the text. At best, that'll make stack traces harder to read.

Reply:
It doesn't affect the _trace_ per se, but I can see how it _might_ be slightly confusing...or not.  E.g.
```
>>> def h():
...   raise Exception("Oops\nSomething broke!")
... 
>>> g()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in g
  File "<stdin>", line 2, in f
  File "<stdin>", line 2, in h
Exception: Oops
Something broke!
```
So, the stack trace itself is intact; the reader just has to figure out that the "Something broke!" line is part of the message which starts with `Exception`.

However, I'm inclined to inconvenience the reader in that way, because I think this is ugly:
```
>>> def h():
...   raise Exception(["Oops","Something broke!"])
... 
>>> g()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in g
  File "<stdin>", line 2, in f
  File "<stdin>", line 2, in h
Exception: ['Oops', 'Something broke!']
```
But, perhaps that's just me.  😛

Reply:
I will try splitting the Error message with `\n` and see how ugly it looks.


Out of the context: I would love to know how do you guys include emoji in Github comments :P.

Reply:
Should I try replacing `\n` with space?

Reply:
> Out of the context: I would love to know how do you guys include emoji in Github comments :P.

When I type `:`, the GitHub UI gives me a popup with emoji selections. Does that not happen for you? It'll refine the emoji search as you type letters, like `:smile` for example. If that's not working for you, I don't know what to say. Maybe clear your browser crash?

> @webbnh has already shown us how ugly that will look :D. Should I try replacing \n with space?

I'm not sure how likely it is that we'll see or care about multi-line output. I was mostly concerned about the fact that, in my testing, even a simple one line message came out with the trailing newline and I'd rather not see that.

The simple solution might be to take Webb's earlier suggestion and just use `{process.stderr.strip()!r}` to remove the trailing noise.

Reply:
[For GitHub emoji, check out the [cheat sheet](https://github.com/ikatyang/emoji-cheat-sheet/blob/master/README.md).]

In a spectacular case, having all the newlines replaced by spaces is likely to be even more annoying than having newlines (because it will remove most of whatever formatting there was in the output, which, if there is a lot of output, would be sad).  I would go with Dave's suggestion of just `strip()`'ing the output.

Reply:
ok sure.